elements/lisk-genesis/src/validate.ts

```diff
@@ -0,0 +1,280 @@
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { Schema } from '@liskhq/lisk-codec';
+import { getAddressFromPublicKey } from '@liskhq/lisk-cryptography';
+import { validator, ErrorObject } from '@liskhq/lisk-validator';
+import { GenesisBlock } from './types';
+import {
+	genesisBlockSchema,
+	genesisBlockHeaderSchema,
+	genesisBlockHeaderAssetSchema,
+	defaultAccountAssetSchema,
+} from './schema';
+import {
+	bufferArrayContains,
+	bufferArraySubtract,
+	bufferArrayLexicographicallyOrdered,
+	bufferArrayContainsSome,
+} from './utils';
+import {
+	EMPTY_BUFFER,
+	GB_GENERATOR_PUBLIC_KEY,
+	GB_REWARD,
+	GB_SIGNATURE,
+	GB_TRANSACTION_ROOT,
+} from './constants';
+import { getHeaderAssetSchemaWithAccountAsset } from './utils/schema';
+
+export const validateGenesisBlock = (
+	block:
+		| GenesisBlock
+		| {
+				// eslint-disable-next-line @typescript-eslint/no-explicit-any
+				[key: string]: any;
+		  },
+	options: { roundLength: number; accountAssetSchema?: Schema },
+): ErrorObject[] => {
+	const { header, payload } = block as GenesisBlock;
+
+	// Genesis block schema validation, only check payload length to be zero
+	const payloadErrors = validator.validate(genesisBlockSchema, {
+		header: EMPTY_BUFFER,
+		payload,
+	});
+
+	// Genesis block header validation
+	const headerErrors = [
+		...validator.validate(genesisBlockHeaderSchema, {
+			...header,
+			asset: EMPTY_BUFFER,
+		}),
+	];
+
+	// Custom header validation not possible with validator
+	if (header.generatorPublicKey !== GB_GENERATOR_PUBLIC_KEY) {
+		headerErrors.push({
+			message: 'should be equal to constant',
+			keyword: 'const',
+			dataPath: 'header.generatorPublicKey',
+			schemaPath: 'properties.generatorPublicKey',
+			params: { allowedValue: GB_GENERATOR_PUBLIC_KEY },
+		});
+	}
+
+	if (header.reward !== GB_REWARD) {
+		headerErrors.push({
+			message: 'should be equal to constant',
+			keyword: 'const',
+			dataPath: 'header.reward',
+			schemaPath: 'properties.reward',
+			params: { allowedValue: GB_REWARD },
+		});
+	}
+
+	if (header.signature !== GB_SIGNATURE) {
+		headerErrors.push({
+			message: 'should be equal to constant',
+			keyword: 'const',
+			dataPath: 'header.signature',
+			schemaPath: 'properties.signature',
+			params: { allowedValue: GB_SIGNATURE },
+		});
+	}
+
+	if (header.transactionRoot !== GB_TRANSACTION_ROOT) {
+		headerErrors.push({
+			message: 'should be equal to constant',
+			keyword: 'const',
+			dataPath: 'header.transactionRoot',
+			schemaPath: 'properties.transactionRoot',
+			params: { allowedValue: GB_TRANSACTION_ROOT },
+		});
+	}
+
+	// Genesis block asset validation
+	const assetSchemaWithAccountAsset = getHeaderAssetSchemaWithAccountAsset(
+		genesisBlockHeaderAssetSchema,
+		options.accountAssetSchema ?? defaultAccountAssetSchema,
+	);
+	const assetErrors = [
```

Comment:
do we need to have this constant here? we are already using spread operator below for `errors`

Reply:
The error concatenation below helps to see what kind of errors were returned. 

```
const errors = [...payloadErrors, ...headerErrors, ...assetErrors];
```